packages/protocol/contracts/governance/DowntimeSlasherIntervals.sol

```diff
@@ -0,0 +1,409 @@
+    returns (bool)
+  {
+    require(signerIndex < numberValidatorsInSet(startBlock), "Bad validator index at start block");
+
+    if (!intervalProofAlreadyCalculated(startBlock, endBlock)) {
+      uint256 accumulatedBitmap = getBitmapForInterval(startBlock, endBlock);
+      return wasDownUsingIntervalBitmap(signerIndex, accumulatedBitmap);
+    }
+
+    return wasDownUsingIntervalProof(startBlock, endBlock, signerIndex);
+  }
+
+  /**
+   * @notice Shows if the user already called the generateProofOfIntervalValidation for
+   * the specific Interval
+   * @param startBlock First block of a calculated downtime Interval.
+   * @param endBlock Last block of the calculated downtime Interval.
+   * @return True if the user already called the generateProofOfIntervalValidation for
+   * the specific Interval
+   */
+  function intervalProofAlreadyCalculated(uint256 startBlock, uint256 endBlock)
+    public
+    view
+    returns (bool)
+  {
+    // It's impossible to have all the validators down in a Interval
+    return userIntervalValidationProof[msg.sender][startBlock][endBlock] != 0;
+  }
+
+  /**
+   * @notice Validates if the signer was down for the specific Interval
+   */
+  function wasDownUsingIntervalProof(uint256 startBlock, uint256 endBlock, uint256 signerIndex)
+    internal
+    view
+    returns (bool)
+  {
+    uint256 intervalBitmap = userIntervalValidationProof[msg.sender][startBlock][endBlock];
+    return wasDownUsingIntervalBitmap(signerIndex, intervalBitmap);
+  }
+
+  /**
+   * @notice Validates if the signerIndex is down in both up interval bitmaps
+   */
+  function wasDownUsingIntervalBitmap(uint256 signerIndex, uint256 intervalBitmap)
+    internal
+    pure
+    returns (bool)
+  {
+    return (intervalBitmap & (1 << signerIndex) == 0);
+  }
+
+  /**
+   * @notice Test if a validator has been down for an specific chain of Intervals.
+   * Requires to:
+   *   - previously called 'generateProofOfIntervalValidation' for every pair
+   * (startIntervals(i), endIntervals(i))
+   *   - startIntervals(0) is the startBlock of the SlashableDowntime
+   *   - endIntervals(i) is included in the interval [startIntervals(i+1) - 1, endIntervals(i+1)]
+   *   - [startBlock, startBlock+SlashableDowntime-1] be covered by
+   * [startIntervals(0), endIntervals(n)]
+   * @param startIntervals List of blocks that starts a previously validated Interval.
+   * startIntervals[0] will be use as the startBlock of the SlashableDowntime
+   * @param endIntervals List of blocks that ends a previously validated Interval.
+   * @param startSignerIndex Index of the signer within the validator set as of the start block.
+   * @param endSignerIndex Index of the signer within the validator set as of the end block.
+   * @return True if the validator signature does not appear in any block within the window.
+   */
+  function wasDownForIntervals(
+    uint256[] memory startIntervals,
+    uint256[] memory endIntervals,
+    uint256 startSignerIndex,
```

Comment:
Consider `signerIndices` as a list here, for more generality.

I think it would actually make the code more readable to *not* have the requirement that `slashableDowntime` be < 1 epoch